File: modules/unity/ios/src/UnityModule.swift
import ExpoModulesCore
import UnityFramework

public class UnityModule: Module, NativeCallsProtocol {
  public static var unityFramework: UnityFramework?
  private var frameworkListener: UnityFrameworkListenerImpl?

  public func definition() -> ModuleDefinition {
    Name("UnityModule")

    // Defines event names that the module can send to JavaScript.
    Events("onUnityMessage", "onUnityUnloaded")

    OnCreate {
      FrameworkLibAPI.registerAPIforNativeCalls(self)
    }

    // Post message to Unity GameObject
    Function("postMessage") { (gameObject: String, methodName: String, message: String) in
      UnityModule.unityFramework?.sendMessageToGO(withName: gameObject, functionName: methodName, message: message)
    }

    // Unload Unity
    Function("unloadUnity") {
      if let ufw = UnityModule.unityFramework {
        ufw.register(self.getFrameworkListener())
        ufw.unloadApplication()
      } else {
        sendEvent("onUnityUnloaded");
        return;
      }
    }

    // Pause/Resume Unity
    Function("pauseUnity") { (pause: Bool) in
      if pause {
        UnityModule.unityFramework?.pause(true)
      } else {
        UnityModule.unityFramework?.pause(false)
      }
    }

    // Enables the module to be used as a native view
    View(UnityView.self) {
    }
  }

  public func sendMessage(toMobileApp message: String) {
    // Send the Unity message directly as an event to React Native
    sendEvent("onUnityMessage", ["message": message])
  }

  // Handle Unity unloaded event
  func handleUnityUnloaded() {
    sendEvent("onUnityUnloaded")
  }

  // Get the framework listener instance
  func getFrameworkListener() -> UnityFrameworkListenerImpl {
    if frameworkListener == nil {
      frameworkListener = UnityFrameworkListenerImpl(module: self)
    }
    return frameworkListener!
  }
}

File: modules/unity/android/src/main/java/com/yousician/unity/UnityModule.kt
package com.yousician.unity

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import com.unity3d.player.UnityPlayer

class UnityModule : Module() {
  companion object {
    private var moduleInstance: UnityModule? = null

    @JvmStatic
    fun sendMessageToMobileApp(message: String) {
      // This method will be called from Unity to send messages to React Native
      moduleInstance?.sendEvent("onUnityMessage", mapOf("message" to message))
    }
  }

  override fun definition() = ModuleDefinition {
    Name("UnityModule")

    // Store module instance for static access
    OnCreate {
      moduleInstance = this@UnityModule
    }

    OnDestroy {
      moduleInstance = null
    }

    // Defines event names that the module can send to JavaScript.
    Events("onUnityMessage", "onUnityUnloaded")

    // Post message to Unity GameObject
    Function("postMessage") { gameObject: String, methodName: String, message: String ->
      UnityPlayer.UnitySendMessage(gameObject, methodName, message)
    }

    // Unload Unity
    Function("unloadUnity") {
      UnityPlayer.quit()
      sendEvent("onUnityUnloaded")
    }

    // Pause/Resume Unity
    Function("pauseUnity") { pause: Boolean ->
      if (pause) {
        UnityPlayer.pause()
      } else {
        UnityPlayer.resume()
      }
    }

    // Simulate window focus change (Android specific)
    Function("windowFocusChanged") { hasFocus: Boolean ->
      UnityPlayer.windowFocusChanged(hasFocus)
    }

    // Enables the module to be used as a native view
    View(UnityView::class) {
      // Prop for keeping player mounted
      Prop("androidKeepPlayerMounted") { view: UnityView, keepMounted: Boolean ->
        view.setKeepPlayerMounted(keepMounted)
      }

      // Prop for full screen mode
      Prop("fullScreen") { view: UnityView, fullScreen: Boolean ->
        view.setFullScreen(fullScreen)
      }
    }
  }
}
